README.md
# Just-In-Time Summoning of Unikernels #

*Jitsu is experimental software. See [LICENSE](#license) for conditions.*

Jitsu is a forwarding DNS server that automatically boots [unikernels](http://wiki.xenproject.org/wiki/Unikernels) on demand. When a DNS query is received, Jitsu first checks for a local unikernel that is mapped to the requested domain. If a unikernel is found, the unikernel is started and its IP is returned to the client. Otherwise, the request is forwarded to the next DNS server. If no DNS requests are received for the unikernel within a given timeout period it is automatically stopped. 

![](docs/jitsu.jpg)

This version of Jitsu has been tested with [MirageOS](https://mirage.io) and [Rumprun](https://github.com/rumpkernel/rumprun) unikernels.

Jitsu supports several backends to manage the unikernel VMs. Currently [libvirt](https://libvirt.org), [XAPI](http://wiki.xenproject.org/wiki/XAPI) and [libxenlight](http://wiki.xen.org/wiki/Choice_of_Toolstacks#Libxenlight_.28libxl.29) are supported. Metadata and internal state is stored in [Irmin](https://github.com/mirage/irmin) and the DNS server is implemented on top of [ocaml-dns](https://github.com/mirage/ocaml-dns).

 - Latest release: https://github.com/mirage/jitsu/releases
 - Bugtracker: https://github.com/mirage/jitsu/issues
 - Development version: https://github.com/mirage/jitsu

[![Build Status](https://travis-ci.org/mirage/jitsu.svg)](https://travis-ci.org/mirage/jitsu)

## Further reading ##
 - [Just-in-Time Summoning of Unikernels (v0.2)](http://www.skjegstad.com/blog/2015/08/17/jitsu-v02/). A more detailed description of how Jitsu works, highlighting some of the new features in v0.2.
 - [Jitsu: Just-In-Time Summoning of Unikernels](https://www.usenix.org/system/files/conference/nsdi15/nsdi15-paper-madhavapeddy.pdf) (by A. Madhavapeddy et al). Research paper describing the first version of Jitsu and how we used it to boot unikernel services in milliseconds on ARM-based [Cubieboards](https://en.wikipedia.org/wiki/Cubieboard).
 - [Towards Heroku for Unikernels: Part 2 - Self Scaling Systems](http://amirchaudhry.com/heroku-for-unikernels-pt2/). Blog post by Amir Chaudhry describing how tools like Jitsu and unikernels can be used to build demand driven clouds.
 - [A unikernel experiment: A VM for every URL](http://www.skjegstad.com/blog/2015/03/25/mirageos-vm-per-url-experiment/). Experimenting with running a separate unikernel per URL.

#### Overview ####

 - [Installing Jitsu](#installing-jitsu)
 - [Getting started](#getting-started)
 - [Examples](#examples)
    - [MirageOS web server](#mirageos-web-server)
    - [nginx in rumprun](#nginx-in-rumprun)
    - [Linux VMs in Virtualbox](#linux-vms-in-virtualbox)
 - [Command line options](#options)
 - [License](#license)

## Installing Jitsu ##

The latest release of Jitsu is available in [`opam`](https://opam.ocaml.org). To install Jitsu, run:

```
$ opam install jitsu
```

When Jitsu is installed it will look for available backends that can be used to start unikernels (or processes). The backends currently supported are `xenctrl` (libxl), `libvirt` and `xen-api-client` (xapi). If a new backend is installed `opam` will reinstall Jitsu to add support for it.

The virtual packages `jitsu-libvirt`, `jitsu-xapi` and `jitsu-libxl` can be used to force Jitsu to be installed with a specific backend. 

To add a backend, either use the virtual package:

```
$ opam install jitsu-libxl
```

Or install the backend directly with `opam`:

```
$ opam install jitsu
$ opam depext xenctrl    # install external dependencies, optional
$ opam install xenctrl
```

To install the latest development version of Jitsu, you can pin Jitsu to the current master branch on Github. This version is *unstable* and changes frequently.

```
$ opam pin add jitsu --dev
$ opam install jitsu
```

If the installation succeeds you should now be able to start Jitsu. If you ran into problems and are using OS X, see below for additional troubleshooting tips.

#### OS X troubleshooting ####
To install the OCaml libvirt bindings on OS X we have to set CPPFLAGS first (due to this [bug](https://github.com/ocaml/opam-repository/issues/2621)). This step can be skipped on other platforms.

```
$ CPPFLAGS="-Wno-error=tautological-compare -Wno-error=unused-function" \
opam install libvirt
```

You should now be able to install Jitsu as usual.

## Getting started ##

Jitsu is initially launched with a list of unikernels, their configurations and a set of parameters that define 

 - how to connect to the virtualization backend
 - how the DNS server should be configured
 - how the unikernels should be managed

A minimal Jitsu configuration could look like this:

```
$ sudo ./jitsu -c xen:/// \
dns=www.openmirage.org,\
ip=192.168.0.22,\
name=mirage-www
```

The command above connects to a local Xen-server (from dom0) through libvirt (the default) and starts the DNS server. Requests for www.openmirage.org will be redirected to the Xen-VM called "mirage-www" with IP 192.168.0.22. If "mirage-www" is not running, Jitsu will start it automatically before responding to the DNS request.

Each unikernel is configured using a set of key/value pairs separated by commas. The parameters that are supported depends on which virtualization backend (libvirt, xapi or libxl) is used to control the unikernels. See [below](#options) or run `./jitsu --help` for a complete set of options.

## Examples ##

### MirageOS web server ###
For this example you need a working MirageOS unikernel with a static IP address and access to dom0 on a Xen server with `libxl` installed. You can also install Xen in Virtualbox, as described [here](http://www.skjegstad.com/blog/2015/01/19/mirageos-xen-virtualbox/) or use one of the [Mirage Vagrant VMs](https://github.com/mirage/mirage-vagrant-vms). An example unikernel that displays the unikernel boot time can be found [here](https://github.com/MagnusS/mirage-stats-demo) (a live version is running [here](http://www.jitsu.v0.no)). Follow the instructions in the README to configure the network settings. After building the unikernel you should have a binary file that can be booted in Xen, usually `mirage-www.xen`. You can also check the generated .xl file and locate the `kernel=` parameter to find the file.

We should now be able to start Jitsu to manage the unikernel:

```
$ sudo jitsu -x libxl \
dns=www.example.org,\
ip=10.0.0.1,\
kernel=mirage-www.xen,\
memory=64000,\
name=www,\
nic=br0
```

This command boots the unikernel when `www.example.org` is resolved. The IP 10.0.0.1 is returned to clients that resolve the domain and the booted unikernel is in `mirage-www.xen`. The VM is given 64MB of memory and access to a network bridge at `br0`. 

If everything worked, Jitsu should now be running a DNS server on localhost. To verify that the domain is automatically started you can run `host` to resolve the domain:

```
host www.example.org 127.0.0.1
Using domain server:
Name: 127.0.0.1
Address: 127.0.0.1#53
Aliases: 

www.example.org has address 10.0.0.1
```

After running `host` you should be able to open the web page on 10.0.0.1 (or telnet to port 80) and ping the IP.

The unikernel is automatically destroyed after the DNS cache entry expires. This timeout can be set with the `--ttl` parameter. See `jitsu --help` for a full list of available parameters and options that can be passed to the `libxl` backend.

### nginx in rumprun ###

[Rumprun unikernels](https://github.com/rumpkernel/wiki/wiki/Tutorial%3A-Getting-Started) can also be managed by Jitsu, but currently only with the `libxl` backend. Prior to running Jitsu, the [rumprun](https://github.com/rumpkernel/rumprun) tool must be used to generate a JSON configuration file that is passed to Jitsu using the `rumprun_config` parameter.

A tutorial for building a unikernel that hosts a static web page in QEMU is available [here](https://github.com/rumpkernel/wiki/wiki/Tutorial%3A-Serve-a-static-website-as-a-Unikernel). To get a Xen unikernel, follow the instructions, but run `rumpbake` with the `xen_pv` parameter instead of `hw_virtio`. After baking the Xen unikernel it must be started once with `rumprun` to obtain the configuration files. To run the Nginx unikernel from the tutorial in Xen you could (depending on your configuration) use a command similar to this:

```
sudo rumprun -T tmp xen -M 64 -i \
    -b images/stubetc.iso,/etc -b images/data.iso,/data \
    -I mynet,xenif,bridge=br0 -W mynet,inet,static,10.0.0.1/24,10.0.1.1 \
    -- nginx.bin -c /data/conf/nginx.conf
```

The command above will boot a rumprun unikernel that mounts two disks (stubetc.iso and data.iso), connect it to the network bridge `br0` and give it the IP 10.0.0.1. The `-T` parameter saves the generated configuration files in the `tmp` directory. You can verify that the unikernel booted correctly by running `sudo xl list`. You can also attach to the console with `sudo xl console [name of unikernel]`. To stop the unikernel, use `sudo xl destroy [name of unikernel]`.

If the unikernel booted correctly we should now be able to use the file `tmp/json.conf` to boot it with Jitsu.

```
jitsu -x libxl dns=www.example.org,\
memory=64000,\
ip=10.0.0.1,\
name=rumprun,\
kernel=nginx.bin,\
disk=images/stubetc.iso@xvda,\
disk=images/data.iso@xvdb,\
nic=br0,\
rumprun_config=tmp/json.cfg 
```

Verify that the unikernel boots in Jitsu by running `host www.example.org 127.0.0.1`. An nginx web server should now be running in a rumprun unikernel on IP 10.0.0.1.

Note that rumprun unikernels currently have to wait for the disk images to be configured by Xen. When the disks are mounted as ISO files (as in this example) the total boot time can be more than a second. An alternative is to use `losetup` to create loopback devices that map to the ISO files. The `-d` parameter can also be used to delay the DNS response to compensate for this. See `jitsu --help` for a full list of available options.

### Linux VMs in Virtualbox ###
Jitsu can be used to control VMs in Virtualbox with libvirt. Note that how well this will work depends on how quickly the VM is able to respond to requests after resuming from suspend (see also the `-d` parameter for how to delay the DNS response).

First, install libvirt and use virsh to display a list of available VMs. Example output:

```
$ virsh list --all
Welcome to virsh, the virtualization interactive terminal.

Type:  'help' for help with commands
       'quit' to quit

virsh # list --all
 Id    Name                           State
----------------------------------------------------
 2     Ubuntu                         running
```

If virsh is unable to connect to Virtualbox, you may have to adjust the connection URI (`-c [uri]`). The default connection URI for Virtualbox is vbox:///session - see [this page](https://libvirt.org/remote.html) for more details. In OS X you may have to set the socket manually, which can be done with 'vbox:///session?socket=path-to-socket'. Remember to use the same connection URI for Jitsu below.

You should now be able to start Jitsu. Use '-m suspend' to set it to suspend the VM on inactivity. Example output:

```
$ sudo ./jitsu dns=www.example.com,ip=127.0.0.1,name=Ubuntu -m suspend -c vbox:///session
Connecting to vbox:///session...
Adding domain 'www.example.com' for VM 'Ubuntu' with ip 127.0.0.1
Adding SOA 'example.com' with ttl=60
Adding A PTR for 'www.example.com' with ttl=60 and ip=127.0.0.1
Starting server on 127.0.0.1:53...
```

To test that Jitsu works, try to resolve the domain with host:

```
$ host www.example.com 127.0.0.1
Using domain server:
Name: 127.0.0.1
Address: 127.0.0.1#53
Aliases:

www.example.com has address 127.0.0.1
```

The domain should now be running.

```
$ virsh dominfo Ubuntu
Id:             2
Name:           Ubuntu
UUID:           6e696eb7-09f4-484c-981b-8d34efa0304d
OS Type:        hvm
State:          running
CPU(s):         3
Max memory:     2147483648 KiB
Used memory:    3166208 KiB
Persistent:     yes
Managed save:   unknown
```

After 2 minutes without DNS requests, Jitsu will suspend the domain automatically. This timeout can be adjusted with the --ttl parameter. Jitsu terminates the VM after 2 x DNS TTL seconds has expired without new requests.

## Options ##
```
    -b ADDR, --bind=ADDR (absent=127.0.0.1)
        Bind local DNS server to interface with this IP

    -c CONNECT, --connect=CONNECT (absent=xen:///)
        Libvirt and Xapi connection string (e.g. xen+ssh://x.x.x.x/system
        or vbox:///session)

    -d SECONDS, --delay=SECONDS (absent=0.1)
        Time to wait in seconds before responding to a DNS query after the
        local VM has started. This delay gives the VM time to open up the
        necessary TCP ports etc. Setting this value too low can result in
        long delays on the first TCP request to the VM.

    -f ADDR, --forwarder=ADDR
        IP address of DNS server queries should be forwarded to if no local
        match is found. Defaults to system default (/etc/resolv.conf) if
        not specified. Set to 0.0.0.0 to disable forwarding.

    --help[=FMT] (default=pager)
        Show this help in format FMT (pager, plain or groff).

    -l PORT, --listen=PORT (absent=53)
        UDP port to listen for DNS queries

    -m MODE, --mode=MODE (absent=destroy)
        How to stop running VMs after timeout. Valid options are suspend,
        destroy and shutdown. Suspended VMs are generally faster to resume,
        but require resources to store state. Note that MirageOS
        suspend/resume is currently not supported on ARM.

    -p PORT, --port=PORT (absent=53)
        UDP port to forward DNS queries to

    --persistdb=path
        Store the Irmin database in the specified path. The default is to
        store the database in memory only. Note that modifying this
        database while Jitsu is running is currently unsupported and may
        crash Jitsu or have other unexpected results.

    --synjitsu=UUID
        UUID of a running Synjitsu compatible unikernel. When specified,
        Jitsu will attempt to connect to a Synjitsu unikernel over Vchan on
        port 'synjitsu' and send notifications with updates on MAC- and
        IP-addresses of booted unikernels. This allows Synjitsu to send
        gratuitous ARP on behalf of booting unikernels and to cache
        incoming SYN packets until they are ready to receive them. This
        feature is experimental and requires a patched MirageOS TCP/IP
        stack.

    -t SECONDS, --ttl=SECONDS (absent=60)
        DNS TTL in seconds. The TTL determines how long the clients may
        cache our DNS response. VMs are terminated after no DNS requests
        have been received for TTL*2 seconds.

    --version
        Show version information.

    -x BACKEND, --backend=BACKEND (absent=libvirt)
        Which backend to use. Currently libvirt, xapi and libxl are
        supported. Xapi and libxl are less tested and should be considered
        experimental.

COMMON CONFIGURATION
    response_delay
        Override default DNS query response delay for this unikernel. See
        also -d.

    wait_for_key
        Wait for this key to appear in Xenstore before responding to the
        DNS query. Sleeps for [response_delay] after the key appears. The
        key should be relative to /local/domain/[domid]/.

    use_synjitsu
        Enable Synjitsu for this domain if not 0 or absent (requires
        Synjitsu support enabled)

LIBVIRT CONFIGURATION
    name
        Name of VM defined in libvirt (ignored if uuid is set)

    uuid
        UUID of VM defined in libvirt (required, but optional if name is
        set)

    dns DNS name (required)

    ip  IP to return in DNS reply (required)

XAPI CONFIGURATION
    name
        Name of VM defined in Xapi (ignored if uuid is set)

    uuid
        UUID of VM defined in Xapi (optional if name is set)

    dns DNS name (required)

    ip  IP to return in DNS reply (required)

LIBXL CONFIGURATION
    name
        Name of created VM (required)

    dns DNS name (required)

    ip  IP to return in DNS reply (required)

    kernel
        VM kernel file name (required)

    memory
        VM memory in bytes (required)

    cmdline
        Extra parameters passed to kernel (optional)

    nic Network device (br0, eth0 etc). Can be set more than once to
        configure multiple NICs (optional)

    script
        Virtual interface (VIF) configuration script. Can be set more than
        once to specify a VIF script per network device (optional)

    disk
        Disk to connect to the Xen VM. Format '[dom0
        device/file]@[hdX/xvdX/sdX etc]'. Can be set more than once to
        configure multiple disks (optional)

    rumprun_config
        Path to file with rumprun unikernel JSON config (optional)
```

## License ##
```
Copyright (c) 2014-2015 Magnus Skjegstad <magnus@skjegstad.com> and contributors.

Permission to use, copy, modify, and distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
```


CHANGES.md
0.3.0 (2015-11-11) 

- Support modular backends. Only available backends are compiled in by default.
- Support xenctrl > 0.9.26
- Support xen-api-client > 0.9.8
- Add virtual packages for jitsu-libvirt, jitsu-libxl, jitsu-xapi
- Update README with new installation instructions

0.2.1 (2015-15-19)

- Add/fix support for multiple DNS entries per VM
- Update storage backend to support latest Irmin API (0.10.0)
- Add VERSION file to track version in builds and releases
- Update README with links to paper and recent blog posts

0.2 (2015-17-08)

General:
- Add support for rumprun unikernels in libxl backend (rumprun_config option)
- New configuration syntax with key/value pairs
- VMs are now destroyed by default (was suspend, can be changed with -m)
- Support for waiting for a key in Xenstore before sending DNS reply (wait_for_key option)
- Support for setting fixed response delay per domain (response_delay option, overrides -d)
- Support more than one DNS name per VM

Storage:
- All storage functionality moved to separate storage module
- Irmin in-memory storage backend to store internal state
- Add --persistdb parameter to persist Irmin db to disk

Synjitsu:
- Add --synjitsu parameter to connect to Synjitsu
- Support for optionally updating synjitsu with the MAC address of newly booted unikernels
- Support for enabling/disabling synjitsu per domain (use_synjitsu option)

VM backends:
- All VM management moved to backend modules that implement backends.mli
- Libvirt support now in separate module
- Add experimental XAPI backend support (from Masoud Koleini @koleini)
- Add experimental libxl backend support (from Dave Scott @djs55)
- Add disk support in libxl backend
- Add support for multiple network interfaces and configuration scripts in libxl backend

Tests:
- Test framework based on alcotest added, but needs more tests
- Implemented tests for options.ml

Bugfixes:
- Destroyed domains can now be restarted (fixes #8)

Packaging:
- Removed version number from opam file in dev repo
- libvirt-dev, libvirt-bin added as depext in Debian and Ubuntu
- conduit, vchan, xen-api-client added as opam dependencies
- Update to ocaml-dns 0.15.3
- Enable Travis on Github repository 

0.1 
- First public release


src/main.ml
(*
 * Copyright (c) 2014-2015 Magnus Skjegstad <magnus@skjegstad.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *)

open Lwt
open Cmdliner

let add_backend_support_msg =
    "Install `libvirt`, `xen-api-client` (xapi) or `xenctrl` (libxl) with opam to add the corresponding backend."

let info =
  let doc =
    "Just-In-Time Summoning of Unikernels. Jitsu is a forwarding DNS server \
     that automatically boots unikernels when their domain is requested. \
     The DNS response is sent to the client after the unikernel has started, \
     enabling the client to use unmodified software to communicate with \
     unikernels that are started on demand. If no DNS requests are received \
     for the unikernel within a given timeout period, the unikernel is automatically \
     stopped." in
  let list_options l =
    List.map (fun x ->
        let (k,v) = x in
        `I ((Printf.sprintf "$(b,%s)" k), v)) l in
  let common_options =
    [ ("response_delay", "Override default DNS query response delay for this unikernel. See also -d.") ;
      ("wait_for_key", "Wait for this key to appear in Xenstore before responding to the DNS query. Sleeps for [response_delay] after the key appears. The key should be relative to /local/domain/[domid]/.") ;
      ("use_synjitsu", "Enable Synjitsu for this domain if not 0 or absent (requires Synjitsu support enabled)") ] in
  let backend_options name = 
      match Vm_backends.lookup name with
      | None -> [] (* Only include options if compiled in *)
      | Some (module M : Backends.VM_BACKEND)-> 
                  [`S (Printf.sprintf "%s CONFIGURATION" (String.uppercase name)) ] @ 
                  (list_options M.get_config_option_list)
  in
  let man =
    [ `S "BACKENDS" ;
      `P "Jitsu can use different backends to control unikernel VMs (or processes). Support for libvirt, xapi and libxl will \
          be enabled if the libraries are available. Note that Xapi and libxl support are less tested and should be considered \
          experimental." ;
      `P add_backend_support_msg ] @
    [ `S "COMMON CONFIGURATION" ] @ (list_options common_options) @
    backend_options "libvirt" @
    backend_options "xapi" @
    backend_options "libxl" @
    [ `S "EXAMPLES";
      `P "$(b,jitsu -c xen:/// -f 8.8.8.8 dns=mirage.io,ip=10.0.0.1,vm=mirage-www)" ;
      `P "Connect to Xen via libvirt. Start unikernel $(b,mirage-www) on requests for $(b,mirage.io) and \
          return IP $(b,10.0.0.1) in DNS. Forward unknown requests to \
          $(b,8.8.8.8).";
      `P "$(b,jitsu -c vbox:///session -m suspend dns=home.local,ip=192.168.0.1,name=ubuntu -t 60)";
      `P "Connect to Virtualbox. Start VM $(b,ubuntu) on requests for $(b,home.local) \
          and return IP $(b,192.168.0.1). Forward unknown requests to system default. \
          Expired VMs are $(b,suspended) after $(b,120) seconds (2 x DNS ttl).";
      `S "AUTHORS";
      `P "Magnus Skjegstad <magnus@skjegstad.com>" ;
      `S "BUGS";
      `P "Submit bug reports to http://github.com/mirage/jitsu";] in
  Term.info "jitsu" ~version:Jitsu_version.current ~doc ~man

let bindaddr =
  let doc = "Bind local DNS server to interface with this IP" in
  Arg.(value & opt string "127.0.0.1" & info ["b"; "bind"] ~docv:"ADDR" ~doc)

let bindport =
  let doc = "UDP port to listen for DNS queries" in
  Arg.(value & opt int 53 & info ["l"; "listen"] ~docv:"PORT" ~doc)

let connstr =
  let doc =
    "Libvirt or Xapi connection string (e.g. xen+ssh://x.x.x.x/system or vbox:///session)"
  in
  Arg.(value & opt string "xen:///" & info ["c"; "connect"] ~docv:"CONNECT" ~doc)

let forwarder =
  let doc =
    "IP address of DNS server queries should be forwarded to if no local match \
     is found. Defaults to system default (/etc/resolv.conf) if not specified. \
     Set to 0.0.0.0 to disable forwarding."
  in
  Arg.(value & opt string "" & info ["f" ; "forwarder"] ~docv:"ADDR" ~doc)

let forwardport =
  let doc = "UDP port to forward DNS queries to" in
  Arg.(value & opt int 53 & info ["p"; "port"] ~docv:"PORT" ~doc)

let response_delay =
  let doc =
    "Time to wait in seconds before responding to a DNS query after the local \
     VM has started. This delay gives the VM time to open up the necessary TCP \
     ports etc. Setting this value too low can result in long delays on the \
     first TCP request to the VM." in
  Arg.(value & opt float 0.1 & info ["d" ; "delay" ] ~docv:"SECONDS" ~doc)

let map_domain =
  let doc =
    "Unikernel configuration. Maps DNS and IP to a unikernel VM. Configuration \
     options are passed as keys and values in the form \"key1=value1,key2=value2...\". \
     A configuration string must be specified for each \
     unikernel Jitsu should control. \
     Required keys are $(b,name), $(b,dns) and $(b,ip). \
     Depending on the selected virtualization backend, additional keys may be supported. \
     See full list of available keys below." in
  Arg.(non_empty & pos_all (array ~sep:',' (t2 ~sep:'=' string string)) [] & info []
         ~docv:"CONFIG" ~doc)

let ttl =
  let doc =
    "DNS TTL in seconds. The TTL determines how long the clients may cache our \
     DNS response. VMs are terminated after no DNS requests have been received \
     for TTL*2 seconds." in
  Arg.(value & opt int 60 & info ["t" ; "ttl" ] ~docv:"SECONDS" ~doc)

let vm_stop_mode =
  let doc =
    "How to stop running VMs after timeout. Valid options are $(b,suspend), \
     $(b,destroy) and $(b,shutdown). Suspended VMs are generally faster to \
     resume, but require resources to store state. Note that MirageOS \
     suspend/resume is currently not supported on ARM." in
  Arg.(value & opt (enum [("destroy" , Vm_stop_mode.Destroy);
                          ("suspend" , Vm_stop_mode.Suspend);
                          ("shutdown", Vm_stop_mode.Shutdown)])
         Vm_stop_mode.Destroy & info ["m" ; "mode" ] ~docv:"MODE" ~doc)

let synjitsu_domain_uuid =
  let doc =
    "UUID of a running Synjitsu compatible unikernel. When specified, \
     Jitsu will attempt to connect to a Synjitsu unikernel over Vchan on port 'synjitsu' \
     and send notifications with updates on MAC- and IP-addresses of booted \
     unikernels. This allows Synjitsu to send gratuitous ARP on behalf of \
     booting unikernels and to cache incoming SYN packets until they are \
     ready to receive them. This feature is $(b,experimental) and requires a patched \
     MirageOS TCP/IP stack."  in
  Arg.(value & opt (some string) None & info ["synjitsu"] ~docv:"UUID" ~doc)

let persistdb =
  let doc =
    "Store the Irmin database in the specified path. The default is to store the database in memory only. \
     Note that modifying this database while Jitsu is running is currently unsupported and may crash Jitsu
     or have other unexpected results." in
  Arg.(value & opt (some string) None & info [ "persistdb" ] ~docv:"path" ~doc)

let log m =
  Printf.fprintf stdout "%s\n%!" m

let or_abort f =
  try f () with
  | Failure m -> (Printf.fprintf stderr "Fatal error: %s" m); exit 1

let or_warn msg f =
  try f () with
  | Failure m -> (log (Printf.sprintf "Warning: %s\nReceived exception: %s" msg m)); ()
  | e -> (log (Printf.sprintf "Warning: Unhandled exception: %s" (Printexc.to_string e))); ()

let or_warn_lwt msg f =
  try_lwt f () with
  | Failure m -> (log (Printf.sprintf "Warning: %s\nReceived exception: %s" msg m)); Lwt.return_unit
  | e -> (log (Printf.sprintf "Warning: Unhandled exception: %s" (Printexc.to_string e))); Lwt.return_unit

let backend =
  let type_if_exists n t =
     match Vm_backends.lookup n with
     | None -> []
     | Some _ -> [(n, t)]
  in
  let backends =
      type_if_exists "libvirt" `Libvirt @
      type_if_exists "libxl" `Libxl @
      type_if_exists "xapi" `Xapi
  in
  let backend_names = 
      String.concat ", " (List.map (fun t -> let n, _ = t in n) backends)
  in
  let default = 
      try
          let _, t = List.hd backends in
          t
      with
      | Failure _ -> raise (Failure (Printf.sprintf "No backend support compiled in. %s" add_backend_support_msg))
  in
  let doc =
    (Printf.sprintf 
     "Which backend to use. This version is compiled with support for: $(b,%s)." backend_names) in
  Arg.(value & opt (enum backends) default & info ["x" ; "backend" ] ~docv:"BACKEND" ~doc)


let jitsu backend connstr bindaddr bindport forwarder forwardport response_delay
    map_domain ttl vm_stop_mode synjitsu_domain_uuid persistdb =
  let (module Vm_backend : Backends.VM_BACKEND) =
    let lookup_or_fail name =
        match Vm_backends.lookup name with
        | None -> raise (Failure (Printf.sprintf "Support for backend '%s' not enabled in this version of jitsu. Install the backend with opam and try again." name))
        | Some (module M : Backends.VM_BACKEND) -> (module M : Backends.VM_BACKEND)
    in
    match backend with
    | `Libvirt -> lookup_or_fail "libvirt"
    | `Xapi -> lookup_or_fail "xapi"
    | `Libxl -> lookup_or_fail "libxl"
  in
  let (module Storage_backend : Backends.STORAGE_BACKEND) = 
    match persistdb with
    | None -> (module Irmin_backend.Make(Irmin_unix.Irmin_git.Memory))
    | Some _ -> (module Irmin_backend.Make(Irmin_unix.Irmin_git.FS))
  in
  let module Jitsu = Jitsu.Make(Vm_backend)(Storage_backend) in
  let rec maintenance_thread t timeout =
    Lwt_unix.sleep timeout >>= fun () ->
    Printf.printf "%s%!" ".";
    or_warn_lwt "Unable to stop expired VMs" (fun () -> Jitsu.stop_expired_vms t) >>= fun () ->
    maintenance_thread t timeout;
  in
  Lwt.async_exception_hook := (fun exn -> log (Printf.sprintf "Exception in async thread: %s" (Printexc.to_string exn)));
  Lwt_main.run (
    ((match forwarder with
        | "" -> Dns_resolver_unix.create () >>= fun r -> (* use resolv.conf *)
          Lwt.return (Some r)
        | "0.0.0.0" -> Lwt.return None
        | _  -> let forwardip = Ipaddr.of_string_exn forwarder in (* use ip from forwarder *)
          let servers = [(forwardip,forwardport)] in
          let config = `Static ( servers , [""] ) in
          Dns_resolver_unix.create ~config:config () >>= fun r ->
          Lwt.return (Some r)
      )
     >>= fun forward_resolver ->
     let connstr = Uri.of_string connstr in
     let synjitsu =
       match synjitsu_domain_uuid with
       | Some s -> Uuidm.of_string s
       | None -> None
     in
     Vm_backend.connect ~connstr () >>= fun r ->
     match r with
     | `Error e -> raise (Failure (Printf.sprintf "Unable to connect to backend: %s" (Jitsu.string_of_error e)))
     | `Ok backend_t ->
       or_abort (fun () -> Jitsu.create backend_t log forward_resolver ~synjitsu ~persistdb ()) >>= fun t ->
       Lwt.pick [(
           (* main thread, DNS server *)
           let add_with_config config_array = (
             let vm_config = (Hashtbl.create (Array.length config_array)) in
             (Array.iter (fun (k,v) -> Hashtbl.add vm_config k v) config_array); (* Use .add to support multiple values per parameter name *)
             let dns_names = Options.get_dns_name_list vm_config "dns" in
             let vm_name = Options.get_str vm_config "name" in
             let vm_ip = Options.get_ipaddr vm_config "ip" in
             let response_delay =  (* override default response_delay if key set in config *)
               match (Options.get_float vm_config "response_delay") with
               | `Error _ -> response_delay
               | `Ok d -> d
             in
             let use_synjitsu =
               match (Options.get_bool vm_config "use_synjitsu"), synjitsu with
               | `Error _, _
               | `Ok _, None -> false (* default to false if use_synjitsu is not set or synjitsu is not enabled *)
               | `Ok v, Some _ -> v
             in
             let wait_for_key =
               match Options.get_str vm_config "wait_for_key" with
               | `Error _ -> None
               | `Ok v -> Some v
             in
             match dns_names, vm_name, vm_ip with
             | `Error e, _, _
             | _, `Error e, _
             | _, _, `Error e -> raise (Failure (Options.string_of_error e))
             | `Ok dns_names, `Ok vm_name, `Ok vm_ip -> begin
                 match (Ipaddr.to_v4 vm_ip) with
                 | None -> raise (Failure (Printf.sprintf "Only IPv4 is supported. %s is not a valid IPv4 address." (Ipaddr.to_string vm_ip)))
                 | Some vm_ip -> begin
                     List.iter (fun dns_name ->
                         log (Printf.sprintf "Adding domain '%s' for VM '%s' with ip %s" (Dns.Name.to_string dns_name) vm_name (Ipaddr.V4.to_string vm_ip)))
                       dns_names;
                     or_abort (fun () -> Jitsu.add_vm t ~dns_names:dns_names ~vm_ip ~vm_stop_mode ~response_delay ~wait_for_key ~use_synjitsu ~dns_ttl:ttl ~vm_config)
                   end
               end
           ) in
           Lwt_list.iter_s add_with_config map_domain >>= fun () ->
           Jitsu.output_stats t () >>= fun () ->
           log (Printf.sprintf "Starting DNS server on %s:%d..." bindaddr bindport);
           try_lwt
             let processor = ((Dns_server.processor_of_process (Jitsu.process t))
                              :> (module Dns_server.PROCESSOR)) in
             Dns_server_unix.serve_with_processor ~address:bindaddr ~port:bindport ~processor
           with
           | e -> log (Printf.sprintf "DNS thread exited unexpectedly with exception: %s" (Printexc.to_string e)); Lwt.return_unit
             >>= fun () ->
             log "DNS server no longer running. Exiting...";
             Lwt.return_unit);

          (* maintenance thread, delay in seconds *)
          (try_lwt
             maintenance_thread t 5.0
           with
           | e -> log (Printf.sprintf "Maintenance thread exited unexpectedly with exception: %s" (Printexc.to_string e)); Lwt.return_unit
             >>= fun () ->
             log "Maintenance thread no longer running. Exiting...";
             Lwt.return_unit)]);
  )

let jitsu_t =
  Term.(pure jitsu $ backend $ connstr $ bindaddr $ bindport $ forwarder $ forwardport
        $ response_delay $ map_domain $ ttl $ vm_stop_mode $ synjitsu_domain_uuid $ persistdb )

let () =
  match Term.eval (jitsu_t, info) with
  | `Error _ -> exit 1
  | _ -> exit 0
